src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,240 @@
+    val outputCollector = LifetimesCollector()
+
+    // extract lifetimes in input argument types
+    fn.valueParameterList?.let { param ->
+        param.selfParameter?.accept(inputCollector)
+        param.valueParameterList.forEach { it.typeReference?.accept(inputCollector) }
+    }
+    if (inputCollector.abort) return false
+
+    // extract lifetimes in output type
+    fn.retType?.typeReference?.accept(outputCollector)
+    if (outputCollector.abort) return false
+
+    val inputLifetimes = inputCollector.lifetimes + lifetimesFromBounds(boundsLifetimes)
+    val outputLifetimes = outputCollector.lifetimes
+
+    if (BodyLifetimeChecker.checkLifetimesUsedInBody(fn.block)) return false
+
+    // check for lifetimes from higher scopes
+    if (inputLifetimes.chain(outputLifetimes).any { !allowedLifetimes.contains(it) }) return false
+
+    // no input lifetimes? easy case!
+    if (inputLifetimes.isEmpty()) return false
+
+    // no output lifetimes, check distinctness of input lifetimes
+    if (outputLifetimes.isEmpty()) {
+        // only unnamed and static, ok
+        if (inputLifetimes.all { it !is ReferenceLifetime.Named }) return false
+        // we have no output reference, so we only need all distinct lifetimes
+        return inputLifetimes.size == countUniqueLifetimes(inputLifetimes)
+    }
+
+    return when {
+        countUniqueLifetimes(outputLifetimes) > 1 -> false
+        inputLifetimes.size == 1 -> {
+            val input = inputLifetimes.first()
+            val output = outputLifetimes.first()
+            when {
+                input is ReferenceLifetime.Named && output is ReferenceLifetime.Named &&
+                    input.name == output.name -> true
+                input is ReferenceLifetime.Named && output == ReferenceLifetime.Unnamed -> true
+                else -> false  // already elided, different named lifetimes or something static going on
+            }
+        }
+        else -> false
+    }
+}
+
+private fun allowedLifetimesFrom(lifetimeParameters: List<RsLifetimeParameter>): Set<ReferenceLifetime> {
+    val allowedLifetimes = HashSet<ReferenceLifetime>()
+    lifetimeParameters
+        .filter { it.bounds.isEmpty() }
+        .mapNotNull { it.name }
+        .map { ReferenceLifetime.Named(it) }
+        .toCollection(allowedLifetimes)
+    allowedLifetimes.add(ReferenceLifetime.Unnamed)
+    allowedLifetimes.add(ReferenceLifetime.Static)
+    return allowedLifetimes
+}
+
+private fun lifetimesFromBounds(bounds: List<RsLifetime>): List<ReferenceLifetime> =
+    bounds.map { it.name }.filterIsInstance<ReferenceLifetime.Named>()
+
+private fun countUniqueLifetimes(lifetimes: List<ReferenceLifetime>): Int = HashSet(lifetimes).size
+
+/**
+ * Are any lifetimes mentioned in the `where` clause? If yes, we don't try to reason about elision.
+ */
+private fun hasWhereLifetimes(whereClause: RsWhereClause?): Boolean {
+    if (whereClause == null) return false
```

Comment:
You can write it like `whereClause ?: return false`, like you did in `checkLifetimesUsedInBody`